orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConfigStageInjectionTransform.java

```diff
@@ -75,17 +72,107 @@ private void replaceStages(PipelineTemplate pipelineTemplate) {
   }
 
   private void injectStages(PipelineTemplate pipelineTemplate) {
-    List<StageDefinition> configStages = templateConfiguration.getStages()
+    // Create initial graph via dependsOn.
+    createDag(pipelineTemplate.getStages());
+
+    // Handle stage injections.
+    injectStages(pipelineTemplate.getStages(), pipelineTemplate.getStages());
+    injectStages(templateConfiguration.getStages(), pipelineTemplate.getStages());
+  }
+
+  private static void createDag(List<StageDefinition> stages) {
+    stages
       .stream()
+      .filter(s -> !s.getDependsOn().isEmpty() && s.getInject() == null)
+      .forEach(s -> s.setRequisiteStageRefIds(
+        stages
+          .stream()
+          .filter(parentCandidate -> s.getDependsOn().contains(parentCandidate.getId()))
+          .map(StageDefinition::getId)
+          .collect(Collectors.toList())
+      ));
+  }
+
+  private static void injectStages(List<StageDefinition> stages, List<StageDefinition> templateStages) {
+    stages.stream()
       .filter(s -> s.getInject() != null)
-      .collect(Collectors.toList());
-    List<StageDefinition> templateStages = pipelineTemplate.getStages()
+      .forEach(s -> {
+        templateStages.add(s);
+
+        InjectionRule rule = s.getInject();
+
+        if (rule.getFirst() != null && rule.getFirst()) {
+          injectFirst(s, templateStages);
+          return;
+        }
+
+        if (rule.getLast() != null && rule.getLast()) {
+          injectLast(s, templateStages);
+          return;
+        }
+
+        if (rule.getBefore() != null) {
+          injectBefore(s, rule.getBefore(), templateStages);
+          return;
+        }
+
+        if (rule.getAfter() != null) {
+          injectAfter(s, rule.getAfter(), templateStages);
+          return;
+        }
+
+        throw new IllegalStateException(String.format("stage did not have any valid injections defined (id: %s)", s.getId()));
+      });
+  }
+
+  private static void injectFirst(StageDefinition stage, List<StageDefinition> allStages) {
+    allStages
       .stream()
-      .filter(s -> s.getInject() != null)
-      .collect(Collectors.toList());
+      .filter(pts -> pts.getRequisiteStageRefIds().isEmpty())
+      .forEach(pts -> pts.getRequisiteStageRefIds().add(stage.getId()));
+  }
 
-    if (!configStages.isEmpty() || !templateStages.isEmpty()) {
-      throw new UnsupportedOperationException("stage injection is not yet supported");
-    }
+  private static void injectLast(StageDefinition stage, List<StageDefinition> allStages) {
+    int numStages = allStages.size();
+    allStages
+      .stream()
+      .filter(s -> !s.getId().equals(stage.getId()))
+      .filter(candidate -> allStages
+        .stream()
+        .filter(s -> !s.getRequisiteStageRefIds().contains(candidate.getId()))
+        .count() == numStages)
```

Comment:
Because I didn't build up an vertex list, so there's no real elegant way to discover relations. As discussed in person, it works - so we could merge it now and make it elegant in a second pass, or close this. Let me know how you'd like to proceed there.